sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs

```diff
@@ -737,8 +737,25 @@ private Block BuildSwitchStatement_CSharp7(SwitchStatementSyntax switchStatement
                     if (CasePatternSwitchLabelSyntaxWrapper.IsInstance(label))
                     {
                         var casePatternSwitchLabel = (CasePatternSwitchLabelSyntaxWrapper)label;
-                        currentSectionBlock = BuildCasePattern(casePatternSwitchLabel,
-                            trueSuccessor: sectionBlock, falseSuccessor: currentSectionBlock);
+
+                        if (ConstantPatternSyntaxWrapper.IsInstance(casePatternSwitchLabel.Pattern))
+                        {
+                            // Creating BranchBlock instead of BinaryBranchBlock will generate some
+                            // false negatives because we will not be able to evaluate the case label
+                            // condition. BinaryBranchBlock causes exceptions in the CSharpExplodedGraph
+                            // because there is no instruction for the branch condition. Such instruction
+                            // would be difficult to generate because it is not present in the AST.
+                            // See https://github.com/SonarSource/sonar-csharp/issues/1801
+                            // and https://github.com/SonarSource/sonar-csharp/issues/1831
+
+                            currentSectionBlock = CreateBranchBlock(casePatternSwitchLabel,
+                                successors: new[] { sectionBlock, currentSectionBlock });

```

Comment:
Create regular BranchBlock for `case Constant when Condition` instead of BinaryBranchBlock because we don't have the required instructions in the CFG to evaluate correctly the `case Constant` part. We are basically falling back to "classic" switch-case handling here.